Part 4 of Aligning Beacon Chain with Latest 2.1 - Bootstrapping Genesis

terenc3t opened the pull request:
Besides validators, we were missing a lot of boot strapping process for genesis state. Such as:

- `indicesForSlots` should be pre-populated with bootstrapped validators using seed 0s
- `crosslinkRecords` should be pre-populated with dynasty number 0 and hash 0s for all 1024 shards
- `totalDeposits` should be pre-populated with total value of bootstrapped validators
- `recentBlockHashes` should be pre-populated with hash 0s for 2 * cycle_count (128 slots)

I rebased from #422 and refactored casperFFG package. The current casperFFG package relies on types package for state and block which introduces circular dependencies. Things like clear attester bitfields, increment justified slot should be done in `blockchain` package, not `casperFFG package`

rawfalafel commented on beacon-chain/types/state.go:
```diff
@@ -33,13 +34,19 @@ func NewActiveState(data *pb.ActiveState) *ActiveState {
 		data: &pb.ActiveState{
 			PendingAttestations: []*pb.AttestationRecord{},
-			RecentBlockHashes:   [][]byte{},
+			RecentBlockHashes:   recentBlockHashes,
```
This seems wasteful. Couldn't validation skip `recentBlockHashes` indexes if `block.slot_number` < `params.CycleLength * 2` instead of allocating a bunch of empty byte arrays? This would mean that the first 128 blocks all have empty arrays.

rawfalafel commented on beacon-chain/types/state.go:
```diff
@@ -48,23 +55,65 @@ func NewGenesisStates() (*ActiveState, *CrystallizedState) {
+
+	// Bootstrap attester indices for slots, each slot contains an array of attester indices.
+	seed := make([]byte, 0, 32)
+	committee, err := casper.ValidatorsByHeightShard(common.BytesToHash(seed), validators, 1, 0)
```
I think this should be `committees`

rawfalafel commented on beacon-chain/types/state.go:
```diff
@@ -48,23 +55,65 @@ func NewGenesisStates() (*ActiveState, *CrystallizedState) {
+	var indexCommittees []uint32
+	for _, committee := range committees {
+		for _, index := range committee.Committee {
+			indexCommittees = append(indexCommittees, uint32(index))
```
Why is `indexCommittees` accumulating the previous committee? Shouldn't there be a reset for every iteration? 

Also, is this casting to `uint` to comply with protobuf? Can we keep the type consistent throughout instead?

rawfalafel commented on beacon-chain/types/state.go:
```diff
@@ -48,23 +55,65 @@ func NewGenesisStates() (*ActiveState, *CrystallizedState) {
+			CrosslinkRecords:       crosslinkRecords,
 			Validators:             validators,
-			IndicesForHeights:      []*pb.ShardAndCommitteeArray{},
+			IndicesForHeights:      indicesForHeight,
```
I'm assuming this should be renamed to `IndicesForSlots`

rawfalafel commented on beacon-chain/types/state.go:
```diff
@@ -48,23 +55,65 @@ func NewGenesisStates() (*ActiveState, *CrystallizedState) {
+		}
+		shardCommittees = append(shardCommittees, c)
+	}
+	indicesForHeight := []*pb.ShardAndCommitteeArray{
```
`indicesForHeights`

rawfalafel commented on beacon-chain/casper/sharding.go:
```diff
@@ -1,23 +1,24 @@
-func ValidatorsByHeightShard(crystallized *types.CrystallizedState) ([]*BeaconCommittee, error) {
-	indices := ActiveValidatorIndices(crystallized)
+func ValidatorsByHeightShard(seed common.Hash, validators []*pb.ValidatorRecord, dynasty uint64, crosslinkStartShard uint64) ([]*BeaconCommittee, error) {
+	indices := ActiveValidatorIndices(validators, dynasty)
```
I think `activeValidators` would be more clear, like the change you made in the casper files.

terenc3t commented:
@rawfalafel thanks for the review, I've addressed all the comments:
- s/indicesForHeights/indicesForSlots
- using uint32 through out for validator indices
- not initializing 128 block hashes of 0s
